### ISSUE TITLE
/implicit/callback route not working after 3.0 update

### ISSUE THREAD

**Issue by sero323** [okta/okta-oidc-js#736]
[FailedLoginRedirection3.0.zip](https://github.com/okta/okta-oidc-js/files/4451421/FailedLoginRedirection3.0.zip)

## I'm submitting this issue for the package(s):

- [ ] jwt-verifier
- [ ] okta-angular
- [ ] oidc-middleware
- [x] okta-react
- [ ] okta-react-native
- [ ] okta-vue

## I'm submitting a:

- [x] Bug report  
- [ ] Feature request
- [ ] Other (Describe below)

## Current behavior

We are using @okta/okta-signin-widget for our login page and to login the user which goes pretty smoothly but once the /implicit/callback route gets called we get a blank page it does not redirect back to the app. With version 2.0.1 it redirected just fine. We are using Authorization Code flow with PKCE. I do see the code and state in the url...
`/implicit/callback#code=Iq9SBFxFfHxkvUqrq67O&state=k2v6ZOeOVq1r1euALfflgxAKLZ5JLFDaKSY7l8IzUgVbVFPAEyykKXOFelIxuyTL`

## Expected behavior

I would expect it to redirect to '/' or the secure route it came from.

## Minimal reproduction of the problem with instructions

See attached for the sample that reproduces the issue. This is straight from the sample-js-react project found [here](https://github.com/okta/samples-js-react/tree/master/okta-hosted-login). The only change is that i switched the following package versions to latest...
`"@okta/okta-react": "^3.0.0",`
`"@okta/okta-signin-widget": "^3.9.0",`

Note: The **testenv** should be updated to set issuer and client Id.

## Environment

- Package Version: 3.0.0
- Browser: Chrome 80
- OS: MacOS
- Node version (`node -v`): v10.19.0
- Other: -

**Comment by Valdermeyder** [comment]
We also experiencing this problem on Chrome, other browsers work fine.
I assume the issue is related to recent Chrome Cookie SameSite Policy is changed in version 80 https://www.okta.com/blog/2019/12/chrome-80-is-on-its-way-and-it-might-change-your-cookies-are-you-ready/ 
However, it is strange that you don't experience the problem using version 2.0.1. It would be nice if somebody from the Okta team will confirm the case.

**Comment by sero323** [comment]
@Valdermeyder i get this issue in Firefox 75.0 as well.

**Comment by sero323** [comment]
Also...from here...https://www.chromium.org/updates/same-site

> April 3, 2020: In light of the extraordinary global circumstances due to COVID-19, we’ve decided to temporarily roll back the enforcement of SameSite cookie labeling on Chrome 80 stable. We recognize the efforts of sites and individual developers who prepared for this change as part of our ongoing effort to improve privacy and security across the web. We appreciate the feedback from across the web ecosystem which has helped inform this decision. We will provide advance notice on here and the Chromium blog when we plan to resume the gradual rollout, which we’re now aiming for over the summer.

**Comment by swiftone** [comment]
@sero323 , @Valdermeyder - I've been able to replicate the issue, I'll see if I can figure out what is going on.  I do not believe it is the SameSite issue, though the timing does make that a logical suspect.

**Comment by swiftone** [comment]
@sero323, @Valdermeyder - The issue looks to be that when you're using the Okta Sign In Widget _AND_ you are using the Authorization Code flow w/PKCE, you need to also set `responseMode` to `query` in the authParams.  ( see https://github.com/okta/okta-signin-widget/issues/1094#issuecomment-602775673 )

We are looking into making the widget smarter about that, but in the meantime I'll update the sample-js-react code. 

Let me know if that does/does not fix your issue.

**Comment by sero323** [comment]
@swiftone that did it. Thank you so much!

**Comment by Valdermeyder** [comment]
@swiftone small clarification we are not using Okta Sign In Widget at all, only okta-react package. And in our case, we don't use pkce (`pkce: false`) flow.
However, when I set `responseMode` to `query` it doesn't fix the problem on Chrome and moreover stops to work on Firefox
Do you have any ideas or other workarounds for the problem?

**Comment by swiftone** [comment]
@Valdermeyder - Thanks for the info, I have a few more questions to help recreate your issue, but first:
- If you aren't using pkce, definitely don't set `responseMode`

Now to my questions:
- Which version of okta-react are you using?  
- Can you confirm in your node_modules which version of `@okta/okta-auth-js` it installed?
- You say you aren't using the widget, are you using the default redirect to login?  (as opposed to using a custom login process)

A common problem is setting your '/' route to redirect to login, and having `/implicit/callback` to handle the post-login, but NOT setting `exact` on your `/` route in react-router.  This means BOTH routes will render for `/implicit/callback`, putting you into a loop.

**Comment by Valdermeyder** [comment]
@swiftone 

> * Which version of okta-react are you using?

version "3.0.0"
> * Can you confirm in your node_modules which version of `@okta/okta-auth-js` it installed?

version "3.0.0" (the same is for "3.0.1")
> * You say you aren't using the widget, are you using the default redirect to login?  (as opposed to using a custom login process)

Yes, we are using default redirect to login. We are providing only these properties to `Security` component:
```javascript
{
  redirectUri: `${window.location.origin}/implicit/callback`,
  pkce: false,
  issuer: '<our value>'
  clientId: '<our value>'
}
```
> 
> A common problem is setting your '/' route to redirect to login, and having `/implicit/callback` to handle the post-login, but NOT setting `exact` on your `/` route in react-router. This means BOTH routes will render for `/implicit/callback`, putting you into a loop.

We are using `exact` fro `/` route and we don't have a loop after login redirect the page stays on `/implicit/callback#id_token=<token>&access_token=<token>&token_type=Bearer&expires_in=3600&scope=openid+profile+email&state=<state_hash>` page forever and no error is shown. However, when debugging `LoginCallback` I am able to see the next error
```javascript
errorCauses: []
errorCode: "INTERNAL"
errorId: "INTERNAL"
errorLink: "INTERNAL"
errorSummary: "Unable to retrieve OAuth redirect params cookie"
message: "Unable to retrieve OAuth redirect params cookie"
name: "AuthSdkError"
```

The difference between Chrome and Safari (where page works fine) is that Safari has next cookies after redirect from Okta
![image](https://user-images.githubusercontent.com/5995043/79013436-02624a80-7b69-11ea-8db8-8e4f291e6195.png)

I hope the provided info will help you to find the reason

**Comment by swiftone** [comment]
@Valdermeyder - Thanks for the detailed breakdown.  Unfortunately, we can't reproduce your problem - Both Chrome and Safari have these cookies in our tests.  What Chrome version and OS are you on?

**Comment by Valdermeyder** [comment]
@swiftone My OS macOS Catalina 10.15.4 (19E287), Chrome Version 81.0.4044.92 (Official Build) (64-bit), Safari Version 13.1 (15609.1.20.111.8)
And one more update, I have discovered that it doesn't work for http://localhost however it works when the application deployed to an https endpoint
Have you tried using http://localhost

**Comment by swiftone** [comment]
@Valdermeyder - Let's summarize as your report is mixed with the original report and there are key differences. 

- You are using okta-react 3.x 
- You are NOT using okta-signin-widget
- You are not using PKCE
- You are seeing this problem only in dev with http://localhost
- The problem is a (hard to see) reported cookie error on /implicit/callback 
- You are seeing this on Chrome and Safari

Correct?

We already test our code on http://localhost, and Chrome has an exception regarding secure cookie errors on localhost, so that should not be the problem.  Have you tried clearing out any stale cookies on http://localhost?

**Comment by SPNexus** [comment]
I just wanted to contribute to this as well.  I'm having the exact same issue and almost the exact same setup as listed above.  I haven't tested moving this out of a localhost environment, as this is an update for an existing app.  
I'm console logging authState from useOktaAuth.  authState.isPending is set to true when I refresh the page, and then it updates to authState.isAuthenticated to false.  I'm not even seeing a redirect to /implicit/callback.  
With the new version of okta-react should we be handling the redirects separately? It was my understanding that that's exactly what /implicit/callback should be taking care of.  
I can also say I have cleared out all cookies, and I'm not seeing anything new generated from my refresh attempts.

**Comment by Valdermeyder** [comment]
@swiftone everything right except the last bullet, doesn't work only on Chrome, works fine on Safari

Yes, I have tried to clean cookies, local storage, etc. 
Have I understand you correctly that you are not able to reproduce the issue on OS macOS Catalina 10.15.4 (19E287), Chrome Version 81.0.4044.92 (Official Build) or on the newer versions of both?

**Comment by nharrisanalyst** [comment]
I would like to add too I am going to open a separate issue 

React-Okat 3.0 gets stuck in a loop if you do not use the exact keyword.

**Comment by dusan-dragon** [comment]
Also for me this does not work:
```
import { ImplicitCallback } from '@okta/okta-react'
..
<Route path="/implicit/callback" component={ImplicitCallback} />
```

instead this works in 3.0.1:
```
import { LoginCallback } from '@okta/okta-react'
..
<Route path="/implicit/callback" component={LoginCallback} />
```

**Comment by swiftone** [comment]
Closing this issue, as the original problem has been corrected with configuration.  Related issues in comments are being tracked by other GH issues.  
(specifically #721 and #760 )

**Comment by Valdermeyder** [comment]
@swiftone what about not working /implicit/callback on Chrome localhost. Would you like me to create a separate issue for that?

**Comment by swiftone** [comment]
@Valdermeyder - Yes please.  We're unable to replicate that, so any details you can share are great.

**Comment by slassen** [comment]
FWIW. For the project, I'm currently upgrading the Okta package on I needed to use the `ImplicitLogin` to `LoginCallback` and explictly add `pkce: false` as a prop to the `Security` component. Hope this helps anyone that might be stuck in the future when upgrading. 

Also, keep in mind that `withAuth` is now `withOktaAuth` if using old HOCs. As well some other functions have had their names changed in the upgraded package.

**Comment by christopherhuii** [comment]
I'm facing a similar issue as OP. I'm using `okta-react` with `okta-signin-widget` with PKCE enabled. In the callback redirect, I'm getting the error `AuthSdkError: Unable to parse a token from the url`.

This happens on `v3.0.1`, but `v.2.0.0` works fine.

```
    <Security
      issuer={process.env.REACT_APP_OKTA_ISSUER}
      clientId={process.env.REACT_APP_OKTA_CLIENT_ID}
      redirectUri={process.env.REACT_APP_OKTA_REDIRECT_URI}
      onAuthRequired={handleAuthRequiredRedirect}
      responseType={['code']}
      pkce
    >
            <Switch>
              <Route path="/" exact component={Home} />
              <SecureRoute path="/protected" component={Protected} />
              <Route path="/implicit/callback" component={LoginCallback} />
            </Switch>
    </Security
```


The setup of my Okta Sign In Widget looks something like this
```
  componentDidMount() {
    this.widget = new OktaSignIn({
      baseUrl: process.env.REACT_APP_OKTA_DOMAIN,
      authParams: {
        pkce: true,
        responseType: 'code',
        responseMode: 'query'
      },
      el: '#sign-in-widget'
    });
    this.widget.showSignInToGetTokens({
      authorizationServerId: process.env.REACT_APP_OKTA_SERVER_ID,
      clientId: process.env.REACT_APP_OKTA_CLIENT_ID,
      redirectUri: process.env.REACT_APP_OKTA_REDIRECT_URI,
      scope: 'openid profile email'
    });
  }
```

**Comment by swiftone** [comment]
@christopherhuii - Have you tried the latest version of the sign-in-widget (4.x)?  It has some updates to help with handling PKCE requests.  If that doesn't work for you, would you mind opening a new issue with these details?  Your symptom looks to be similar to the OP here, but the cause looks to be different.

**Comment by christopherhuii** [comment]
@swiftone Thanks for the prompt response. I tried updating the signin-widget to the latest version, but that brought up additional issues. I can try to document them more clearly in a new issue.

I know this is the wrong repository, but the main difference I saw for the signin-widget, is:
```
    this.widget.showSignInToGetTokens({
      authorizationServerId: process.env.REACT_APP_OKTA_SERVER_ID,
      clientId: process.env.REACT_APP_OKTA_CLIENT_ID,
      redirectUri: process.env.REACT_APP_OKTA_REDIRECT_URI,
      scope: 'openid profile email'
    });
```
https://github.com/okta/okta-signin-widget#showsignintogettokens
The `authorizationServerId` is no longer in the documentation for `showSignInToGetTokens`. I'm not sure if that's a result of the version upgrade or if added that parameter accidentally a while back. I think that might be related to the signin-widget issue I experienced when upgrading.
